clients/drcachesim/scheduler/scheduler.cpp

```diff
@@ -1135,6 +1160,65 @@ scheduler_tmpl_t<RecordType, ReaderType>::check_and_fix_modulo_problem_in_schedu
     return STATUS_SUCCESS;
 }
 
+template <typename RecordType, typename ReaderType>
+typename scheduler_tmpl_t<RecordType, ReaderType>::scheduler_status_t
+scheduler_tmpl_t<RecordType, ReaderType>::read_switch_sequences()
+{
+    std::unique_ptr<ReaderType> reader, reader_end;
+    if (!options_.kernel_switch_trace_path.empty()) {
+        reader = get_reader(options_.kernel_switch_trace_path, verbosity_);
+        if (!reader || !reader->init()) {
+            error_string_ +=
+                "Failed to open kernel switch file " + options_.kernel_switch_trace_path;
+            return STATUS_ERROR_FILE_OPEN_FAILED;
+        }
+        reader_end = get_default_reader();
+    } else if (!options_.kernel_switch_reader) {
+        // No switch data provided.
+        return STATUS_SUCCESS;
+    } else {
+        if (!options_.kernel_switch_reader_end) {
+            error_string_ += "Provided kernel switch reader but no end";
+            return STATUS_ERROR_INVALID_PARAMETER;
+        }
+        reader = std::move(options_.kernel_switch_reader);
+        reader_end = std::move(options_.kernel_switch_reader_end);
+        if (!reader->init()) {
```

Comment:
The issue is that reader_t::init can block, and it does block for ipc_reader_t.  That means it can't be called until we're ready to consume the first output.  This is how the other readers passed to the scheduler are handled as well.

Reply:
Added a comment.